lib/Driver/ToolChains.cpp

```diff
@@ -1109,25 +1109,47 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
     break;
   case LinkKind::DynamicLibrary:
     Arguments.push_back("-shared");
-    if (getTriple().getOS() == llvm::Triple::Linux) {
-      if (getTriple().getSubArch() == llvm::Triple::SubArchType::ARMSubArch_v7 ||
-          getTriple().getSubArch() == llvm::Triple::SubArchType::ARMSubArch_v6) {
-        Arguments.push_back("-Wl,-Bsymbolic");
-      }
+    switch (getTriple().getArch()) {
+    default:
+      break;
+    case llvm::Triple::arm:
+    case llvm::Triple::armeb:
+    case llvm::Triple::thumb:
+    case llvm::Triple::thumbeb:
+      // Avoid emitting R_ARM_REL32 which is not supported by shared objects
+      // on ELF targets.
+      Arguments.push_back("-Wl,-Bsymbolic");
+      break;
     }
     break;
   }
 
-  addPrimaryInputsOfType(Arguments, context.Inputs, types::TY_Object);
-  addInputsOfType(Arguments, context.InputActions, types::TY_Object);
+  // Select the linker to use
+  llvm::SmallString<128> Linker;
 
-  context.Args.AddAllArgs(Arguments, options::OPT_Xlinker);
-  context.Args.AddAllArgs(Arguments, options::OPT_linker_option_Group);
-  context.Args.AddAllArgs(Arguments, options::OPT_F);
+  switch (getTriple().getArch()) {
+  default:
+    break;
+  case llvm::Triple::arm:
+  case llvm::Triple::armeb:
+  case llvm::Triple::thumb:
+  case llvm::Triple::thumbeb:
+    // BFD linker has issues wrt relocation of the protocol conformance
+    // section on these targets, it also generates COPY relocations for
+    // final executables, as such, unless specified, we default to gold
+    // linker.
+    if (!context.Args.hasArg(options::OPT_use_ld)) {
+      Linker = "gold";
+    }
+    break;
+  }
 
-  if (!context.OI.SDKPath.empty()) {
-    Arguments.push_back("--sysroot");
-    Arguments.push_back(context.Args.MakeArgString(context.OI.SDKPath));
+  if (const Arg *A = context.Args.getLastArg(options::OPT_use_ld)) {
```

Comment:
The other comment got lost, but the switch should go in the else branch here, and the SmallString can be replaced by StringRef.